File: te_depth.py
#!/usr/bin/env python3
"""Compute T&E (Trial & Error) depth for proving non-3-colorability of sudoku patterns.

T&E depth measures the minimum nesting depth of tentative color assignments
needed to prove a pattern's induced graph is not 3-colorable:

  T&E(0): Constraint propagation alone (singleton forced assignments) reaches contradiction.
  T&E(n): By tentatively assigning a color to a vertex and showing contradiction
           via T&E(n-1), we can eliminate that color. Repeating at each nesting
           level eventually eliminates all possible 3-colorings.

Propagation = unit propagation: when a vertex has exactly one color left,
assign it and remove that color from all neighbors. Repeat until fixpoint.
"""

import sys
import time
from pathlib import Path


def build_sudoku_neighbors():
    """Build adjacency lists for the 81-cell sudoku graph."""
    nbrs = [[] for _ in range(81)]
    for v in range(81):
        r, c = divmod(v, 9)
        br, bc = 3 * (r // 3), 3 * (c // 3)
        for u in range(81):
            if u == v:
                continue
            ru, cu = divmod(u, 9)
            if ru == r or cu == c or (3 * (ru // 3) == br and 3 * (cu // 3) == bc):
                nbrs[v].append(u)
    return nbrs


SUDOKU_NBRS = build_sudoku_neighbors()


def bitstring_to_cells(bs):
    return [i for i, ch in enumerate(bs.strip()) if ch == '1']


def build_local_adj(cells):
    """Build local adjacency lists for the induced subgraph on `cells`."""
    idx = {c: i for i, c in enumerate(cells)}
    n = len(cells)
    adj = [[] for _ in range(n)]
    for i, c in enumerate(cells):
        for nb in SUDOKU_NBRS[c]:
            if nb in idx:
                adj[i].append(idx[nb])
    return adj


# Domain bitmask: bit 0 = color 1, bit 1 = color 2, bit 2 = color 3
# So domain 7 = 0b111 = all three colors available

_POPCOUNT = [bin(i).count('1') for i in range(8)]


def propagate(doms, adj):
    """Unit propagation for 3-coloring. Modifies `doms` in place.

    When a vertex has exactly one color left (singleton), remove that color
    from all its neighbors. Repeat until no more singletons to process.

    Returns True if a contradiction is found (some vertex has 0 colors).
    """
    n = len(doms)
    # Seed queue with existing singletons
    queue = [v for v in range(n) if doms[v] and (doms[v] & (doms[v] - 1)) == 0]

    while queue:
        v = queue.pop()
        d = doms[v]
        if d == 0:
            return True
        if d & (d - 1):  # no longer singleton (should not happen, but safe)
            continue
        for u in adj[v]:
            if doms[u] & d:
                doms[u] &= ~d
                if doms[u] == 0:
                    return True
                if (doms[u] & (doms[u] - 1)) == 0:  # became singleton
                    queue.append(u)
    return False


def te_solve(doms, adj, depth):
    """Test whether non-3-colorability can be proved at T&E nesting depth `depth`.

    At the current level we repeatedly:
      1. Pick any (vertex, color) where the vertex still has multiple colors.
      2. Tentatively assign that color and recurse at depth-1.
      3. If recursion proves contradiction, eliminate that color and propagate.
      4. Repeat until either a contradiction is reached or no more eliminations
         are possible.

    Returns True if non-3-colorability is proved.
    """
    doms = list(doms)  # work on a copy
    if propagate(doms, adj):
        return True
    if depth == 0:
        return False

    n = len(doms)
    changed = True
    while changed:
        changed = False
        # Try vertices in MRV order (smallest multi-valued domain first)
        order = sorted(range(n), key=lambda v: _POPCOUNT[doms[v]] if _POPCOUNT[doms[v]] > 1 else 99)

        for v in order:
            if _POPCOUNT[doms[v]] <= 1:
                break  # remaining are singletons or already assigned

            for c in (1, 2, 4):
                if not (doms[v] & c):
                    continue
                trial = list(doms)
                trial[v] = c
                if te_solve(trial, adj, depth - 1):
                    # Color c proved impossible for v — eliminate it
                    doms[v] &= ~c
                    changed = True
                    if doms[v] == 0:
                        return True  # all colors eliminated → contradiction
                    if propagate(doms, adj):
                        return True

            if changed:
                break  # restart scan with updated state

    return False


def compute_te_depth(cells, max_depth=5):
    """Compute the T&E depth of the induced subgraph on `cells`."""
    adj = build_local_adj(cells)
    n = len(adj)
    doms = [7] * n
    for d in range(max_depth + 1):
        if te_solve(list(doms), adj, d):
            return d
    return -1


def main():
    import argparse

    parser = argparse.ArgumentParser(description='Compute T&E depth for sudoku patterns')
    parser.add_argument('input', help='File with 81-char bitstrings, one per line')
    parser.add_argument('-d', '--max-depth', type=int, default=5, help='Maximum T&E depth to check (default: 5)')
    parser.add_argument('-o', '--output', help='Output file (one depth per line). Defaults to stdout.')
    parser.add_argument('--single', type=int, metavar='IDX', help='Process only the pattern at this 1-based index')
    args = parser.parse_args()

    lines = [l.strip() for l in Path(args.input).read_text().splitlines() if l.strip()]
    total = len(lines)

    if args.single:
        idx = args.single - 1
        cells = bitstring_to_cells(lines[idx])
        d = compute_te_depth(cells, args.max_depth)
        print(f"Pattern {args.single}/{total}: T&E({d})")
        return

    print(f"Processing {total} patterns (max_depth={args.max_depth})...", file=sys.stderr)

    depth_counts = {}
    results = []
    t0 = time.time()

    for i, bs in enumerate(lines):
        cells = bitstring_to_cells(bs)
        d = compute_te_depth(cells, args.max_depth)
        results.append(d)
        depth_counts[d] = depth_counts.get(d, 0) + 1

        if (i + 1) % 100 == 0 or (i + 1) == total:
            elapsed = time.time() - t0
            rate = (i + 1) / elapsed if elapsed > 0 else 0
            print(f"  {i + 1}/{total} ({elapsed:.1f}s, {rate:.1f}/s)", file=sys.stderr)

    elapsed = time.time() - t0
    print(f"\nDone in {elapsed:.1f}s", file=sys.stderr)
    print(f"Distribution:", file=sys.stderr)
    for d in sorted(depth_counts.keys()):
        label = f"T&E({d})" if d >= 0 else "UNKNOWN"
        print(f"  {label}: {depth_counts[d]}", file=sys.stderr)

    # Write results
    out = open(args.output, 'w') if args.output else sys.stdout
    for d in results:
        out.write(f"{d}\n")
    if args.output:
        out.close()
        print(f"Wrote {len(results)} results to {args.output}", file=sys.stderr)


if __name__ == '__main__':
    main()

File: rust/src/te_depth.rs
/// T&E (Trial & Error) depth computation for 3-coloring of small graphs.
///
/// T&E depth measures the minimum nesting depth of tentative color assignments
/// needed to prove a graph is not 3-colorable:
///
///   T&E(0): Constraint propagation alone reaches contradiction.
///   T&E(n): By tentatively assigning colors and showing contradiction via
///           T&E(n-1), we can eliminate colors. Repeating at each nesting
///           level eventually eliminates all possible 3-colorings.
///
/// Propagation includes:
///   - Naked singles: vertex with one color left → remove from neighbors
///   - Triangle all-different: in any 3-clique with 3 colors, each vertex
///     gets a distinct color. This gives hidden singles and naked pairs.

use crate::sudoku_graph;

/// Domain bitmask: bit 0 = color 0, bit 1 = color 1, bit 2 = color 2.
/// Full domain = 0b111 = 7.
type Domain = u8;

const FULL: Domain = 7;

#[inline]
fn dom_count(d: Domain) -> u32 {
    (d & 7).count_ones()
}

/// Unit propagation for 3-coloring domains.
/// When a vertex has exactly one color left (singleton), remove that color
/// from all neighbors. Repeat until fixpoint.
/// Returns true if a contradiction (empty domain) is found.
#[inline]
fn propagate_singles(doms: &mut [Domain], adj: &[u16]) -> bool {
    let n = doms.len();
    let mut queue: u16 = 0;
    for v in 0..n {
        if doms[v] != 0 && dom_count(doms[v]) == 1 {
            queue |= 1 << v;
        }
    }

    while queue != 0 {
        let v = queue.trailing_zeros() as usize;
        queue &= queue - 1;
        let d = doms[v];
        if d == 0 {
            return true;
        }
        if dom_count(d) != 1 {
            continue;
        }
        let mut nbrs = adj[v];
        while nbrs != 0 {
            let u = nbrs.trailing_zeros() as usize;
            nbrs &= nbrs - 1;
            if doms[u] & d != 0 {
                doms[u] &= !d;
                if doms[u] == 0 {
                    return true;
                }
                if dom_count(doms[u]) == 1 {
                    queue |= 1 << u;
                }
            }
        }
    }
    false
}

/// Full constraint propagation: naked singles + triangle all-different.
/// Loops until joint fixpoint.
/// Returns true if a contradiction is found.
fn propagate(doms: &mut [Domain], adj: &[u16], triangles: &[[usize; 3]]) -> bool {
    loop {
        if propagate_singles(doms, adj) {
            return true;
        }

        let mut changed = false;
        for tri in triangles {
            let [a, b, c] = *tri;

            // Hidden singles: if a color can only go in one vertex → assign
            for color in [1u8, 2, 4] {
                let in_a = (doms[a] & color) != 0;
                let in_b = (doms[b] & color) != 0;
                let in_c = (doms[c] & color) != 0;
                let count = in_a as u8 + in_b as u8 + in_c as u8;

                if count == 0 {
                    return true; // no vertex can take this color → contradiction
                }
                if count == 1 {
                    let v = if in_a { a } else if in_b { b } else { c };
                    if dom_count(doms[v]) > 1 {
                        doms[v] = color;
                        changed = true;
                    }
                }
            }

            // Naked pairs: two vertices share same 2-element domain → third gets remainder
            let verts = [a, b, c];
            for i in 0..3 {
                let j = (i + 1) % 3;
                let k = (i + 2) % 3;
                let vi = verts[i];
                let vj = verts[j];
                let vk = verts[k];
                if dom_count(doms[vi]) == 2 && doms[vi] == doms[vj] {
                    let remainder = FULL & !doms[vi];
                    let new_dom = doms[vk] & remainder;
                    if new_dom != doms[vk] {
                        doms[vk] = new_dom;
                        changed = true;
                        if new_dom == 0 {
                            return true;
                        }
                    }
                }
            }
        }

        if !changed {
            return false;
        }
    }
}

/// Find all triangles in the induced subgraph.
fn find_triangles(adj: &[u16]) -> Vec<[usize; 3]> {
    let n = adj.len();
    let mut tris = Vec::new();
    for i in 0..n {
        for j in (i + 1)..n {
            if adj[i] & (1 << j) == 0 {
                continue;
            }
            // k must be adjacent to both i and j
            let common = adj[i] & adj[j];
            let mut mask = common >> (j + 1);
            while mask != 0 {
                let bit = mask.trailing_zeros() as usize;
                tris.push([i, j, j + 1 + bit]);
                mask &= mask - 1;
            }
        }
    }
    tris
}

/// Test whether non-3-colorability can be proved at T&E nesting depth `depth`.
fn te_solve(doms: &mut [Domain], adj: &[u16], triangles: &[[usize; 3]], depth: u32) -> bool {
    if propagate(doms, adj, triangles) {
        return true;
    }
    if depth == 0 {
        return false;
    }

    let n = doms.len();
    loop {
        let mut changed = false;

        for v in 0..n {
            if dom_count(doms[v]) <= 1 {
                continue;
            }

            for color in [1u8, 2, 4] {
                if doms[v] & color == 0 {
                    continue;
                }
                let mut trial: [Domain; 16] = [0; 16];
                trial[..n].copy_from_slice(&doms[..n]);
                trial[v] = color;
                if te_solve(&mut trial[..n], adj, triangles, depth - 1) {
                    doms[v] &= !color;
                    changed = true;
                    if doms[v] == 0 {
                        return true;
                    }
                    if propagate(doms, adj, triangles) {
                        return true;
                    }
                }
            }
        }

        if !changed {
            return false;
        }
    }
}

/// Compute the T&E depth of the induced subgraph on the given cells.
/// Returns the minimum depth d such that T&E(d) proves non-3-colorability.
/// Returns -1 if max_depth is exceeded.
pub fn compute_te_depth(cells: &[u8], max_depth: u32) -> i32 {
    let adj = sudoku_graph::induced_subgraph(cells);
    let n = adj.len();
    let triangles = find_triangles(&adj);
    let base_doms = [FULL; 16];

    for d in 0..=max_depth {
        let mut doms = [0u8; 16];
        doms[..n].copy_from_slice(&base_doms[..n]);
        if te_solve(&mut doms[..n], &adj, &triangles, d) {
            return d as i32;
        }
    }
    -1
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_triangle_is_3colorable() {
        let adj: Vec<u16> = vec![0b110, 0b101, 0b011];
        let triangles = find_triangles(&adj);
        let mut doms = [FULL; 3];
        assert!(!te_solve(&mut doms, &adj, &triangles, 5));
    }

    #[test]
    fn test_k4_needs_te1() {
        let adj: Vec<u16> = vec![0b1110, 0b1101, 0b1011, 0b0111];
        let triangles = find_triangles(&adj);
        let n = 4;
        // T&E(0): all domains full, no propagation triggers.
        let mut doms0 = [FULL; 4];
        assert!(!te_solve(&mut doms0[..n], &adj, &triangles, 0));
        // T&E(1): try a color, propagation cascades to contradiction.
        let mut doms1 = [FULL; 4];
        assert!(te_solve(&mut doms1[..n], &adj, &triangles, 1));
    }
}
